### PR TITLE
Nfdiv 1555

### PR DESCRIPTION
**Pull request by velimd**
### JIRA link ###
https://tools.hmcts.net/jira/browse/NFDIV-1555

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/routes.ts [hunk 1]
```diff
@@ -60,12 +65,23 @@ export class Routes {
     app.post(DOCUMENT_MANAGER, handleUploads.array('files[]', 5), errorHandler(documentManagerController.post));
     app.get(`${DOCUMENT_MANAGER}/delete/:index`, errorHandler(documentManagerController.delete));
 
+    const isRouteForUser = (req: AppRequest, res: Response, next: NextFunction): void => {
+      const isApp2Route = [APPLICANT_2, RESPONDENT].some(prefixUrl => req.path.includes(prefixUrl));
+      if ((isApp2Route && !req.session.isApplicant2) || (!isApp2Route && req.session.isApplicant2)) {
+        return res.redirect('/');
```
Comment (linusnorton) [inline src/main/routes.ts:71]: infinite death loop?

File: src/main/routes.ts [hunk 2]
```diff
@@ -81,6 +97,12 @@ export class Routes {
       errorHandler(new AccessCodePostController(applicant2AccessCodeContent.form.fields).post)
     );
 
+    app.get(SWITCH_TO_SOLE_APPLICATION, errorHandler(new SwitchToSoleApplicationGetController().get));
+    app.post(
+      SWITCH_TO_SOLE_APPLICATION,
+      errorHandler(new SwitchToSoleApplicationPostController(switchToSoleAppContent.form.fields).post)
+    );
```
Comment (linusnorton) [inline src/main/routes.ts:104]: Why were these two needed?

Comment (linusnorton) [reply src/main/routes.ts:104]: I see

File: src/main/steps/switch-to-sole-application/get.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { GetController } from '../../app/controller/GetController';
+
+import { generateContent } from './content';
+
+export class SwitchToSoleApplicationGetController extends GetController {
```
Comment (linusnorton) [inline src/main/steps/switch-to-sole-application/get.ts:5]: why is this needed?

File: src/test/functional/features/applicant-middleware.feature [hunk 4]
```diff
@@ -0,0 +1,21 @@
+Feature: Homepage
+
+  Background:
```
Comment (linusnorton) [inline src/test/functional/features/applicant-middleware.feature:3]: feels like this should be a nightly test

File: src/main/routes.ts (hunk 1 above)
Comment (velimd) [reply src/main/routes.ts:71]: dont know what that is, sounds like an anime?

File: src/main/steps/switch-to-sole-application/get.ts (hunk 3 above)
Comment (velimd) [reply src/main/steps/switch-to-sole-application/get.ts:5]: i had to remove this from app1 sequence as app2 also uses the page, i guess i could have just used a normal get controller but i noticed every other page outside of the applicant routes were uses custom controllers that did the same thing.

File: src/test/functional/features/applicant-middleware.feature (hunk 4 above)
Comment (velimd) [reply src/test/functional/features/applicant-middleware.feature:3]: that can be arranged

File: src/main/routes.ts (hunk 1 above)
Comment (linusnorton) [reply src/main/routes.ts:71]: what if you get redirected to / and that in turn redirects you to /applicant2/ something, which triggers this and redirects you back to / - it might be best to redirect to the error page.

**Review by linusnorton — APPROVED** [review]

**Review by lara-mpt — APPROVED** [review]
